### PR TITLE
[pcl/python] Implemented python program-gen for PCL components

### PR DESCRIPTION
**Pull request by Zaid-Ajaj**
# Description

Implements python program-gen for components

Still todo:

 - [x] worked out correct types for component arguments
 - [x] fixed component scope traversals
 - [x] made object-typed component variables optional by default

Fixes # (issue)

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/codegen/pcl/utilities.go [hunk 1]
```diff
@@ -236,3 +236,13 @@ func GenerateMultiArguments(
 		}
 	}
 }
+
+func SortedKeys[V any](properties map[string]V) []string {
+	propertyNames := make([]string, 0)
```
Comment (Frassle) [inline pkg/codegen/pcl/utilities.go:241]: I'd rephrase the names in this to not reference properties. so s/properties/m s/propertyNames/keys etc

File: pkg/codegen/pcl/utilities.go [hunk 2]
```diff
@@ -236,3 +236,13 @@ func GenerateMultiArguments(
 		}
 	}
 }
+
+func SortedKeys[V any](properties map[string]V) []string {
```
Comment (Frassle) [inline pkg/codegen/pcl/utilities.go:240]: SortedStringKeys.

**Review by Frassle — APPROVED** [review]
Looks reasonable. I guess we're gonna have to do something smarter about optional/non-optional at some point.

**Comment by Zaid-Ajaj** [comment]
bors merge
